### ISSUE TITLE
The breakpoint in python functions between PyGILState_Ensure and PyGILState_Release is never hit.

### ISSUE THREAD

**Issue by YasinDemirkan** [Microsoft/PTVS#3003]
I've discovered that breakpoints in python functions between PyGILState_Ensure and PyGILState_Release if located in a separated thread is never hit. The following test environment can be used to investigate the problem. 

First you have to initialize and build the Python Interpreter and then call the python function "task" from a thread. Here is my code for this purpose:

// test.cpp : Defines the entry point for the console application.
//
#include "stdafx.h"
#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
#ifdef _DEBUG
#undef _DEBUG
#include <python.h>
#define _DEBUG
#else
#include <python.h>
#endif
#include <numpy/ndarrayobject.h>
#include <string>
#include <thread>

static const int     ERROR = 0;
static const int     OK = 1;

PyThreadState  *g_MainThreadState;
PyObject       *m_pModule    = 0x0;
std::string     m_ModulePath = "mymodule";

void CallPyFunction(const char *function);

int main()
{
	std::string m_SysCommand;

	// 1) Initialise python interpretator
	if (!Py_IsInitialized()) {
		Py_Initialize();
		Py_AtExit(Py_Finalize);
	}

	// 2) Initialise python thread mechanism 
	if (!PyEval_ThreadsInitialized()) {
		PyEval_InitThreads();
	}
	PyEval_SaveThread();

	PyGILState_STATE gstate = PyGILState_Ensure();

	// 3) Save the main thread state
	g_MainThreadState = PyThreadState_Get();

	// 4) create a fake command line directly after Py_Initialize() (Tensorflow can't be imported without this)
	wchar_t const *dummy_args[] = { L"Python", NULL };  // const is needed because literals must not be modified
	wchar_t const **argv = dummy_args;
	int             argc = sizeof(dummy_args) / sizeof(dummy_args[0]) - 1;
	PySys_SetArgv(argc, const_cast<wchar_t **>(argv)); // const_cast allowed, because PySys_SetArgv doesn't change argv
	if (PyErr_Occurred()) {
		return false;
	}

	// 5) Load main module (without reload)
	PyObject  *pName = PyUnicode_FromString(m_ModulePath.c_str());
	m_pModule = PyImport_Import(pName);
	Py_DECREF(pName);

	PyGILState_Release(gstate);

	///////////////// Python Function Call from a Thread ////////////////////
	std::thread worker([]
	{
		PyGILState_STATE gstate = PyGILState_Ensure();

		CallPyFunction("task");

		PyGILState_Release(gstate);
	});
	worker.join();

	/////////////////////////////////////////////////////////////////////////

    return 0;
}

void CallPyFunction(const char *function)
{
	PyObject *pFunc, *pArgs, *pValue;
	pFunc = PyObject_GetAttrString(m_pModule, function);
	if (pFunc && PyCallable_Check(pFunc)) {
		pArgs = PyTuple_New(0);
		pValue = PyObject_CallObject(pFunc, pArgs);
		if (pArgs) {
			Py_XDECREF(pArgs);
		}
		if (pFunc) {
			Py_XDECREF(pFunc);
		}
		if (pValue) {
			Py_XDECREF(pValue);
		}
	}
}

In addition, a test python file "mymodule.py" is required, which looks like this:

// mymodule.py : Defines python test script file
//
import ptvsd   

ptvsd.enable_attach(secret = 'mysecret', address = ('127.0.0.1', 8080))
ptvsd.wait_for_attach()

txt = 2.3

def task( ):
    global val
    val = 4.9       // <---------- Breakpoint must be set here
   return

The breakpoint, if set in the function "task", is never reached.

If the full visual studio project is needed to this test environment, I can provide it.

**Comment by int19h** [comment]
This is because you start the thread from C++. There's a general problem with threads due to the fact that Python doesn't offer a generic callback to indicate that a new thread has been initialized for Python use, where we could set our trace hooks via `sys.settrace`. For normal Python threads, we do it by monkey-patching the `thread` module. But when the threads are created through other means, we're out of the loop. 

That said, I think we do have an existing way to explicitly register a new thread with the debugger - try doing this:
```py
import ptvsd.debugger
...
def task():
    ptvsd.debugger.new_external_thread() # registers current thread with ptvsd
    ...
```
Note that this isn't meant to be a public API at the moment, and I'm not 100% sure that it would work, but it seems like it ought to.

If this is an acceptable workaround, then on our side, we should work on making this a proper, documented and stable public API, exported directly from `ptvsd` rather than a submodule.

@zooba, @brettcannon - is this something we could raise with Python core as a helpful extensibility point? I think if we had a way to register a callback to be invoked when `PyGILState_Ensure` is used for the first time on a new thread, we could reliably detect new Python threads. It's worth noting that PyDevD also does monkey patching to detect threads and set trace hooks for them - they do it for more cases (e.g. they monkey patch `QThread` as well), but it would still fail on code like this. So this appears to be a rather generic piece of missing functionality.

**Comment by zooba** [comment]
This would probably be a notification on `PyThreadState_New`, rather than `PyGILState_Ensure`, but it's feasible.

It might be interesting to investigate simply hooking the exported function though, since that would work for a wider range of versions :)

**Comment by int19h** [comment]
Sure, but it would also be limited to CPython on Win32. And we'd need to take a dependency on something like MinHook.

**Comment by zooba** [comment]
Other interpreters don't have `PyThreadState_New`, so we'd need separate approaches for them anyway.

**Comment by int19h** [comment]
Isn't that an argument in favor of a higher-level callback? Other interpreters don't have `PyThreadState_New`, but they do have some concept of "new Python-aware thread has been started somehow, and is about to start executing Python code" (normally I'd expect that to happen whenever they initialize their thread-local structures - which would be the storage backing `settrace`, at the minimum). So presumably that would be where they would invoke said callback, letting us do `settrace`, and then it'd all magically just work.

**Comment by YasinDemirkan** [comment]
@int19h: I have made the changes you offered me but it is still not working. The new python test file "mymodule.py" looks like this:

```python
import ptvsd   
import ptvsd.debugger

ptvsd.enable_attach(secret = 'mysecret', address = ('127.0.0.1', 8080))
ptvsd.wait_for_attach()

val = 2.3

def task( ):
    global val
    ptvsd.debugger.new_external_thread()      # registers current thread with ptvsd
    val = 4.9                                 # <--------- Breakpoint is set here
    return
```


By the way I'm using ptvsd version 3.2.0, Visual Studio version 15.3 and Python version 3.5.4(all 64-Bit Versions).

**Comment by int19h** [comment]
Ah, I think I see the problem. We will call `settrace`, but it'll only apply to any *new* frames from `new_external_thread` onwards, not to the one where you call it from. To make this work properly, we'd need a new function with some CPython-specific magic.

In the meantime, the only other workaround I can think of is to start the thread using standard Python facilities - i.e. `thread` or `threading`, instead of using the C++ standard library; e.g. using `PyRun_...`. The newly created thread can immediately invoke into C++, and then from there back to Python, and tracing (and hence breakpoint) will work in the latter.

**Comment by YasinDemirkan** [comment]
This is only a test environment. The real project is much bigger than that and it is completely written in C/C++ with lots of threads working in parallel. Changing the software architecture because of debugging would be unreasonable.

OK... then I have to wait for newer versions of ptvsd. Maybe you guys solve the problem sometime.

**Comment by zooba** [comment]
You should be able to refactor that code only slightly to make it work though.

Before:
```python
def task( ):
    global val
    val = 4.9
    return
```

After:
```python
def task( ):
    ptvsd.debugger.new_external_thread()
    return _task()

def _task( ):
    global val
    val = 4.9
    return
```

Or alternatively you could make your launcher code call a different function than `task`. You could even import `ptvsd` and call the function from C code if that is easiest (probably not :) ):

```c
PyObject *mod = PyImport_ImportModule("ptvsd.debugger");
if (!mod)
    return FAILURE;
if (!PyObject_CallMethod(mod, "new_external_thread", NULL)) {
    Py_DECREF(mod);
    return FAILURE;
}
Py_DECREF(mod);

CallPyFunction("task")
```

**Comment by YasinDemirkan** [comment]
I do not know why your first suggestion works, but it works. I have not tested your second suggestion because I do not want to change the C/C++ code for debugging, but it maybe interesting to other users. 

Thank you and int19h for the support.

**Comment by zooba** [comment]
It works because the trace function takes effect on the next _call_ after you enable it. By putting the code you want to debug into a separate function, it means we can handle it properly - this is basically just a quirk of Python's built in debugging functionality.

I'd like to keep this open so we can come up with a nicer way to support this, but as the workaround is pretty easy we'll leave it done :)

**Comment by int19h** [comment]
Should we at least move `new_external_thread` to `ptvsd` itself, and add a docstring explaining what it does, and the caveats?

**Comment by zooba** [comment]
Can we have a generic `ptvsd.set_trace()` that handles all the cases? The consistency would be nice to have with `pdb`, and if the instruction is "call set_trace on any new thread and it will (open localhost socket/enable tracing/nop) depending on program state" that would be neat.

**Comment by zooba** [comment]
See https://www.python.org/dev/peps/pep-0553/ - if `ptvsd.set_trace()` also breaks immediately then we have a drop-in replacement that will work cleanly, and it's easy to have a flag to not block immediately.

**Comment by int19h** [comment]
I don't see why not - it would be a trivial wrapper around existing functions.

BTW, I took a look at the implementation of `set_trace` in PDB, and it actually walks the frame stack backwards setting `f_trace`, not just calling `sys.settrace` - so we can actually do this in pure Python as well, and then debugging would work in the same frame that called the function.

**Comment by pauldotknopf** [comment]
I am using visual studio code, which as far as I can tell, requires exactly version 3.0.0 of ptvsd.

In 3.0.0, I don't see the new ```new_external_thread``` method.

What would you recommend I do? I am having similar problems with PyOtherSide and it's background worker thread.
